src/Controller/FacebookAuthController.php

```diff
@@ -145,32 +168,63 @@ public function returnFromFb() {
       return $this->redirect('user.login');
     }
 
-    $this->facebookManager->setClient($facebook)->authenticate();
+    $state = $this->persistentDataHandler->get('oAuth2State');
 
-    // Checks that user authorized our app to access user's email address.
-    if (!$this->facebookManager->checkPermission('email')) {
-      drupal_set_message($this->t('Facebook login failed. This site requires permission to get your email address from Facebook. Please try again.'), 'error');
-      $this->persistentDataHandler->set('reprompt', TRUE);
+    if (empty($_GET['state']) || ($_GET['state'] !== $state)) {
+      unset($_SESSION['oauth2state']);
+      drupal_set_message($this->t('Facebook login failed. Unvalid oAuth2 State.'), 'error');
       return $this->redirect('user.login');
     }
 
+    $this->facebookManager->setClient($facebook)->authenticate();
+
     // Gets user's FB profile from Facebook API.
     if (!$fb_profile = $this->facebookManager->getUserInfo()) {
       drupal_set_message($this->t('Facebook login failed, could not load Facebook profile. Contact site administrator.'), 'error');
       return $this->redirect('user.login');
     }
 
     // Gets user's email from the FB profile.
-    if (!$email = $this->facebookManager->getEmail($fb_profile)) {
+    if (!$email = $this->facebookManager->getUserInfo()->getEmail()) {
       drupal_set_message($this->t('Facebook login failed. This site requires permission to get your email address.'), 'error');
       return $this->redirect('user.login');
     }
 
-    // Saves access token to session so that event subscribers can call FB API.
+    $data = [];
+
+    $data_points = explode(',', $this->getDataPoints());
+
+    foreach ($data_points as $data_point) {
```

Comment:
Not the best way. What happens if I need birthday? Would I get an error?

What you should check instead is if you can get the email address and name with the current scopes